### PR TITLE
Add cloud logging link for TPU

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Feature request for adding cloud logging link for TPU (b/322535956):

Context: The TPU logs in old infra are from **GKE** workload. After some exploration, I noticed 2 ways to get **GCE** TPU logs in cloud logging. 1) get all tpu worker instance IDs and have OR operation in query; 2) work with cloud ops agent, and write all logs into a file. 

I was not able to find the instance ID from the python lib [here](https://cloud.google.com/python/docs/reference/tpu/latest/google.cloud.tpu_v1.types.Node), so go ahead with 2nd strategy.

* Add helper functions to install `google-cloud-ops-agent` 
* Write stderr and output into a log file
* Update JAX Jsonnet file to avoid error from `END_SCRIPT`

# Tests

Please describe the tests that you ran on Cloud VM to verify changes.

**Instruction and/or command lines to reproduce your tests:** ...
Upload to Airflow and test with log URL ([here](https://screenshot.googleplex.com/A3vzNuUZUYQ5NND))

**List links for your tests (use go/shortn-gen for any internal link):** ...
* log link for a multihost test in flax: [here](https://cloudlogging.app.goo.gl/AdHx3wcCTzEzYbWj6)
* log link for a test with startup script: [here](https://cloudlogging.app.goo.gl/MDjJZGUbtyXPhYTw5)
* log link for a PT Jsonnet test: [here](https://cloudlogging.app.goo.gl/T4cwYzQNt46qwJMD6)
* log link for a JAX Jsonnet test: [here](https://cloudlogging.app.goo.gl/QePDyaSBuMDBFKrL7)

Unfortunately, this seems not working with TensorFlow. I got incompatible error below for both single host and multihosts.

```
Feb  5 05:16:05 t1v-n-9a1d2687-w-0 google_metadata_script_runner[3979]: startup-script: [2024-02-05T05:16:05+0000] Could not refresh the google-cloud-ops-agent apt repositories.
Feb  5 05:16:05 t1v-n-9a1d2687-w-0 google_metadata_script_runner[3979]: startup-script: Please check your network connectivity and make sure you are running a supported
Feb  5 05:16:05 t1v-n-9a1d2687-w-0 google_metadata_script_runner[3979]: startup-script: ubuntu distribution. See https://cloud.google.com/stackdriver/docs/solutions/ops-agent/#supported_operating_systems
Feb  5 05:16:05 t1v-n-9a1d2687-w-0 google_metadata_script_runner[3979]: startup-script: for a list of supported platforms.
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run one-shot tests and provided workload links above if applicable. 
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Comment by RissyRan** [comment]
Hi @will-cromar, what will be our next step for this PR? At least this will benefit for TPU on JAX & PT via GCE platform

**Review by will-cromar — COMMENTED** [review]
How do the logs created by the ops agent differ from the ones already recorded by Composer? Any output we stream back is recorded with the `projects/cloud-ml-auto-solutions/logs/airflow-worker` type.

```
resource.labels.environment_name="ml-automation-solutions"
logName="projects/cloud-ml-auto-solutions/logs/airflow-worker"
labels."task-id"="flax_gpt2_oscar-v4-32.run_model"
```

Is it just that the ops agent logs are tagged with the instance name?

File: xlml/utils/tpu.py [hunk 1]
```diff
@@ -140,6 +143,9 @@ def create_queued_resource_request(tpu_name: str, ssh_keys: ssh.SshKeys) -> str:
     logging.info('Create QR response: {}'.format(response))
     # TODO(wcromar): do anything about failures
 
+    log_url = f'https://console.cloud.google.com/logs/query;query=resource.type%3D%22gce_instance%22%20AND%0Alog_id%2528%22{tpu_name}%22%2529;?project={gcp.project_name}'
```
Comment (will-cromar) [inline xlml/utils/tpu.py:146]: For formatting complicated HTTP query strings like this, you can use `urlencode`: https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode

That lets you build a more natural multi-line string in Python for the Cloud logging query before it gets mangled.

File: xlml/utils/tpu.py [hunk 2]
```diff
@@ -315,4 +321,8 @@ def ssh_tpu(
           )
       },
   )
-  ssh_group.run(cmds)
+
+  queued_resource_name = qualified_name.split('/')[-1]
+  # write stderr and output into a log file
+  run_cmds = f'set -euxo pipefail; ({cmds}) 2>&1 | tee -a /tmp/{queued_resource_name}'
```
Comment (will-cromar) [inline xlml/utils/tpu.py:327]: Is this shell `bash`? `|&` combine both stderr and stdout https://stackoverflow.com/a/35385038

File: xlml/utils/tpu.py [hunk 3]
```diff
@@ -315,4 +321,8 @@ def ssh_tpu(
           )
       },
   )
-  ssh_group.run(cmds)
+
+  queued_resource_name = qualified_name.split('/')[-1]
```
Comment (will-cromar) [inline xlml/utils/tpu.py:325]: nit: prefer `os.path.basename` for the last segment of a path-like string

**Review by chandrasekhard2 — APPROVED** [review]

File: xlml/utils/tpu.py (hunk 1 above)
Comment (RissyRan) [reply xlml/utils/tpu.py:146]: Sounds good. Will take a look

**Comment by RissyRan** [comment]
> How do the logs created by the ops agent differ from the ones already recorded by Composer? Any output we stream back is recorded with the `projects/cloud-ml-auto-solutions/logs/airflow-worker` type.
> 
> ```
> resource.labels.environment_name="ml-automation-solutions"
> logName="projects/cloud-ml-auto-solutions/logs/airflow-worker"
> labels."task-id"="flax_gpt2_oscar-v4-32.run_model"
> ```
> 
> Is it just that the ops agent logs are tagged with the instance name?

Yeah, I think the benefit of this is to filter logs to a single host for debugging ([one example](https://screenshot.googleplex.com/B27E6hvRuHZYhNJ)). This will benefit multihost/multislice testing. 

I think we could add a feature to populate both log links. I can add the one you provided to a future PR (if we want).

**Pull request closed without merging**